### PR TITLE
Fix food cap issues

### LINKED ISSUE #74: When bands adds members, option of food is listed as a paid extra
_From @pelololoquin on June 8, 2016 15:38_

image

_Copied from original issue: magfest/ubersystem#1848_

### PR DESCRIPTION
**Pull request by binary1230**
- create better distinction between those who paid for food and those who get it free automatically
- update reg UI to show when this is happening

Fixes #74.

### CODE REVIEW

**Comment by binary1230** [comment]
IMPORTANT NOTE: this changes the meaning of the c.FOOD_STOCK cap. (recall that when we reach the cap for food, currently 130, no one else is allowed to buy food and the options are hidden)

before the cap on food was calculated like this:
cap = num of people paying for food + performers + performer comps + staff badges
number of meals we have to make = cap

now the cap is calculated:
cap = num of people paying for food ONLY.
number of meals we have to make = cap + performers + performer comps + staff badges

TLDR: we need to drop the cap once we've deployed this from 130 to around 80 in order to make sure we don't overcommit to the number of meals

File: magstock/__init__.py [hunk 1]
```diff
@@ -57,7 +65,7 @@ def food_cost(self):
 
     @property
     def auto_food(self):
-        return self.badge_type in [c.STAFF_BADGE, c.GUEST_BADGE]
+        return self.badge_type in [c.STAFF_BADGE, c.GUEST_BADGE] or hasattr(self, "auto_food_extra") and self.auto_food_extra
```
Comment (kitsuta) [inline magstock/__init__.py]: Rather than do the `auto_food_extra` attribute, why not just do the `self.group and self.group.band is not None` check in here?

Comment (binary1230) [inline magstock/__init__.py]: so this "auto_food_extra" is implemented by the band plugin, and you can see that here: https://github.com/magfest/bands/pull/43/files

basically the band plugin just says that any Attendee who is a member of a Group that is a Band gets free food.

Comment (binary1230) [inline magstock/__init__.py]: it would be cool if the band plugin could actually override this function as if it were a derived class and then call the base function, but this was the best way I could figure out how to do it.

File: magstock/templates/regextra.html [hunk 2]
```diff
@@ -60,14 +60,18 @@
 </script>
```
Comment (binary1230) [inline magstock/templates/regextra.html]: This section of if statements are starting to get pretty hairy, at some future point we should simplify it.

I tested the hell out of this with the food cap not hit, with it hit, and with combinations of guest badges, staff badges, attendee badges in band groups, etc, and it seemed to work just fine.

File: magstock/__init__.py (hunk 1 above)
Comment (binary1230) [inline magstock/__init__.py]: @kitsuta the reason is that the magstock plugin probably shouldn't know anything about bands.  so this is it setting up a hook where other plugins can say which things should get free food.  In that sense, the band plugin is modifying the magstock plugin, and the magstock plugin doesn't have to know anything about bands in order to work.

Either way is probably Ok since magstock will probably use the band plugin forever from now on, I just figured it would be good to maintain a one way interface between these two plugins where bands plugin can mess with magstock stuff, but magstock plugin doesn't have to know anything about bands

**Comment by kitsuta** [comment]
+1 besides my comment

File: magstock/__init__.py [hunk 3]
```diff
@@ -34,9 +34,17 @@ def PREREG_DONATION_OPTS(self):
...
+        """
+        return [a for a in self.query(Attendee).all() if a.gets_food]
+
     def food_purchasers(self):
-        return self.query(Attendee).filter(or_(Attendee.purchased_food == True,
-                                               Attendee.badge_type.in_([c.STAFF_BADGE, c.GUEST_BADGE])))
+        """
+        :return: list of anyone who paid money to get food
+        """
+        return self.query(Attendee).filter(Attendee.purchased_food == True)
```
Comment (binary1230) [inline magstock/__init__.py]: note that the meaning of food_purchasers has now changed.  it used to include anyone getting food (free or not), now includes ONLY people who paid for it.

File: magstock/__init__.py [hunk 4]
```diff
@@ -34,9 +34,17 @@ def PREREG_DONATION_OPTS(self):
 
 @Session.model_mixin
 class SessionMixin:
+    def food_consumers(self):
+        """
+        :return: list of anyone who gets food, regardless of if they're comp'd food or if they paid for it
+        """
+        return [a for a in self.query(Attendee).all() if a.gets_food]
```
Comment (binary1230) [inline magstock/__init__.py]: btw this slightly sucks that in order to use a DB query here, we have to rewrite a bunch of the existing logic for ```.gets_food``` as sqlalchemy statements.  since the amount of attendees in magstock will be less than 600 and this   is for a not often used report, I took the dumb/easy/lame way out and just do the filtering in python.

File: magstock/__init__.py (hunk 1 above)
Comment (kitsuta) [inline magstock/__init__.py]: Yeah, er, why shouldn't the magstock plugin know anything about bands? Event plugins are the "last in line" and can have as many dependencies as we want.

There SHOULD be a one-way interface - magstock should know stuff about bands, bands shouldn't know anything about magstock. Unless we expect to use auto_food_extra in multiple plugins (which right now we don't), you're adding a property to the band plugin that won't be used unless we have the magstock plugin running alongside it - it doesn't make that much sense to do that. Like, imagine someone installs the band plugin in their event. They see the `auto_food_extra` property and they go looking around for where it's used in the band plugin... except, it isn't used, leaving the poor coder confused.

If it wasn't so dependent on event-specific business logic - or if the concept of 'free food' was integral to bands - it'd be a different story.

Comment (binary1230) [inline magstock/__init__.py]: Hm you know what, yea you're right.  I was remember the (unrelated) band agreement template relationship backwards.  I'll switch this.

**Comment by binary1230** [comment]
fixes https://github.com/magfest/magstock/issues/74

**Comment by kitsuta** [comment]
Psst, if you comment "fixes X" then that won't actually do anything when you merge the issue. I updated the description.

+1
